extmod/modgzip.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * This file is part of the CircuitPython project
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018 Josh Klar (https://github.com/klardotsh)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#pragma GCC diagnostic ignored "-Wcast-align"
+
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+
+#include "py/runtime.h"
+#include "py/binary.h"
+#include "extmod/modgzip.h"
+#include "extmod/moduzlib.h"
+
+#define MAGIC 0x1F8B
+#define DEFLATED 0x08
+#define FTEXT 0x01
+#define FMULTI 0x02
+#define FEXTRA 0x04
+#define FNAME 0x08
+#define FCOMMENT 0x10
+
+#define DEBUG_ENABLE 0
+
+#if DEBUG_ENABLE // print debugging info
+#define DEBUG_printf DEBUG_printf
+#else // don't print debugging info
+#define DEBUG_printf(...) (void)0
+#endif
+
+mp_obj_t mod_gzip_decompress(size_t n_args, const mp_obj_t *args) {
+    mp_obj_t data = args[0];
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(data, &bufinfo, MP_BUFFER_READ);
+
+    // Read, and then promptly discard, the GZIP header. This is mostly a
+    // sanity-checking step, consistent with CPython's lack of exposing this
+    // information anywhere (which feels like an odd oversight, and some third
+    // parties seem to agree: the pure-Python gzip implementation "gzippy"
+    // explicitly has a Header class which can be read from)
+
```

Comment:
Can you use uzlib_gzip_parse_header() to do the parsing?

Reply:
Had I known that function existed, yep, probably could have. It'd require a bit more tinkering to `moduzlib.c` to rebase to that, but... seems sane to me. Looks like they do the same "discard most everything" logic I do.

Reply:
This refactor was WAY easier than I thought and the final implementation is _significantly_ cleaner, thank you for this!